packages/contracts-bedrock/src/universal/OptimismMintableERC20FactoryInterop.sol

```diff
@@ -0,0 +1,107 @@
+    bytes32 internal constant HASH_ONION_SLOT = 0xae706b81046b35e941cb9737b644ec489aa4690fffd2c3f501ccbed8fd5b0400;
+
+    /// @notice Initial layer of the hash onion.
+    bytes32 internal constant INITIAL_ONION_LAYER = keccak256(abi.encode(0));
+
+    /// @notice Semantic version.
+    /// @custom:semver +interop-beta.1
+    function version() public view override returns (string memory) {
+        return string.concat(super.version(), "+interop-beta.1");
+    }
+
+    /// @notice Getter for the hashOnion.
+    /// @return hashOnion_ The hashOnion value.
+    function hashOnion() public view returns (bytes32 hashOnion_) {
+        assembly {
+            hashOnion_ := sload(HASH_ONION_SLOT)
+        }
+    }
+
+    /// @notice Verifies the hashOnion and stores the provided addresses.
+    ///         Stores deployments of the provided tokens if the computed hashOnion is valid. It can be considered
+    ///         complete when the hashOnion matches the initial layer.
+    ///         To submit the proof, the caller must provide the local and remote tokens in the same order as how the
+    ///         hashOnion was computed, starting from the chosen inner layer. To verify and store all the
+    ///         deployments, the caller must start with the INITIAL_ONION_LAYER.
+    /// @param _localTokens  Array of local token addresses.
+    /// @param _remoteTokens Array of remote token addresses.
+    /// @param _startingInnerLayer The starting inner layer of the hashOnion.
+    function verifyAndStore(
+        address[] calldata _localTokens,
+        address[] calldata _remoteTokens,
+        bytes32 _startingInnerLayer
+    )
+        external
+    {
+        bytes32 _currentHashOnion = hashOnion();
+        if (_currentHashOnion == INITIAL_ONION_LAYER) revert DeploymentsAlreadyStored();
+        if (_localTokens.length != _remoteTokens.length) revert TokensLengthMismatch();
+
+        // Unpeel the hash onion and store the deployments
+        bytes32 innerLayer = _startingInnerLayer;
+        for (uint256 i; i < _localTokens.length; i++) {
+            innerLayer = keccak256(abi.encodePacked(innerLayer, abi.encodePacked(_localTokens[i], _remoteTokens[i])));
+            deployments[_localTokens[i]] = _remoteTokens[i];
+            emit DeploymentStored(_localTokens[i], _remoteTokens[i]);
+        }
+
+        if (innerLayer != _currentHashOnion) revert InvalidProof();
+        assembly {
+            sstore(HASH_ONION_SLOT, _startingInnerLayer)
+        }
+
+        emit HashOnionUpdated(_startingInnerLayer, _currentHashOnion);
+    }
+
+    /// @notice One-time setter for the hashOnion value to be called by the ProxyAdmin.
+    /// @param _hashOnion The new hashOnion value.
+    function setHashOnion(bytes32 _hashOnion) external {
+        // TODO: Research if this is the correct address to check for.
+        if (msg.sender != Ownable(Predeploys.PROXY_ADMIN).owner()) revert Unauthorized();
```

Comment:
We have been researching with gotzen and to have the correct sender from L1 we found that it has to be aliased.

`AddressAliasHelper.undoL1ToL2Alias(msg.sender)`

Reply:
Perfect, implemented it :+1